cdap-common/src/main/java/co/cask/cdap/common/UnauthorizedException.java

```diff
@@ -16,21 +16,39 @@
 
 package co.cask.cdap.common;
 
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+
 /**
  * Thrown when a user is not authorized to perform an operation.
  */
 public class UnauthorizedException extends Exception {
 
+  private final boolean authenticationError;
+
   public UnauthorizedException() {
     super();
+    this.authenticationError = true;
   }
 
   public UnauthorizedException(String msg, Throwable throwable) {
     super(msg, throwable);
+    this.authenticationError = true;
   }
 
   public UnauthorizedException(String message) {
     super(message);
+    this.authenticationError = true;
```

Comment:
everywhere this constructor is called, its because of an authentication error. For these scenarios, per HTTP standard, we need to throw a 401 - which is an unauthorized operation whose reason is an authentication error. 

When authentication is successful, but the Principal is unauthorized to perform an operation on an entity, we throw 403, because the reason for the unauthorized response is other than authentication error.